src/Functions/UserDefined/UserDefinedSQLObjectsLoaderFromZooKeeper.cpp

```diff
@@ -0,0 +1,416 @@
+        else
+            return false;
+    }
+
+    LOG_DEBUG(log, "Object {} removed", backQuote(object_name));
+    return true;
+}
+
+bool UserDefinedSQLObjectsLoaderFromZooKeeper::getObjectDataAndSetWatch(
+    const zkutil::ZooKeeperPtr & zookeeper,
+    String & data,
+    const String & path,
+    UserDefinedSQLObjectType object_type,
+    const String & object_name)
+{
+    const auto object_watcher = [watch_queue = watch_queue, object_type, object_name](const Coordination::WatchResponse & response)
+    {
+        if (response.type == Coordination::Event::CHANGED) [[maybe_unused]]
+            bool push_result = watch_queue->emplace(object_type, object_name);
+        /// NOTE: Event::DELETED is processed as child event by getChildren watch
+    };
+
+    Coordination::Stat entity_stat;
+    String object_create_query;
+    return zookeeper->tryGetWatch(path, data, &entity_stat, object_watcher);
+}
+
+ASTPtr UserDefinedSQLObjectsLoaderFromZooKeeper::parseObjectData(const String & object_data, UserDefinedSQLObjectType object_type)
+{
+    switch (object_type)
+    {
+        case UserDefinedSQLObjectType::Function: {
+            ParserCreateFunctionQuery parser;
+            ASTPtr ast = parseQuery(
+                parser,
+                object_data.data(),
+                object_data.data() + object_data.size(),
+                "",
+                0,
+                global_context->getSettingsRef().max_parser_depth);
+            return ast;
+        }
+    }
+    UNREACHABLE();
+}
+
+ASTPtr UserDefinedSQLObjectsLoaderFromZooKeeper::tryLoadObject(
+    const zkutil::ZooKeeperPtr & zookeeper, UserDefinedSQLObjectType object_type, const String & object_name)
+{
+    String path = getNodePath(zookeeper_path, object_type, object_name);
+    LOG_DEBUG(log, "Loading user defined object {} from zk path {}", backQuote(object_name), path);
+
+    try
+    {
+        String object_data;
+        bool exists = getObjectDataAndSetWatch(zookeeper, object_data, path, object_type, object_name);
+
+        if (!exists)
+        {
+            LOG_INFO(log, "User-defined object '{}' can't be loaded from path {}, because it doesn't exist", backQuote(object_name), path);
+            return nullptr;
+        }
+
+        return parseObjectData(object_data, object_type);
+    }
+    catch (...)
+    {
+        tryLogCurrentException(log, fmt::format("while loading user defined SQL object {}", backQuote(object_name)));
+        return nullptr; /// Failed to load this sql object, will ignore it
+    }
+}
+
+Strings UserDefinedSQLObjectsLoaderFromZooKeeper::getObjectNamesAndSetWatch(
+    const zkutil::ZooKeeperPtr & zookeeper, UserDefinedSQLObjectType object_type)
+{
+    auto object_list_watcher = [watch_queue = watch_queue, object_type](const Coordination::WatchResponse &)
+    { [[maybe_unused]] bool push_result = watch_queue->emplace(object_type, ""); };
```

Comment:
You can just ignore the emplace result or use the `[[maybe_unused]]` attribute without creating a temporary variable.

Reply:
Unfortunatelly `emplace` is marked [[nodiscard]] attribute. As for [[maybe_unused]], it can be used only in declarations (as variable in this case), but not function calling.

Reply:
`push_result` is not a very clear name. Let's call it `inserted`, and let's check it with chassert

Reply:
I agree. Did it